Introduce GetPods() to replace GetKubeletDockerContainers()

yifan-gu opened the pull request:
A follow-up fix for #5526
To smooth the review, I only replaced in the syncPod/SyncPods. There are other places needed to be cleanup. Let's do it in following PRs.

PTAL thank you! @dchen1107 @vmarmol

Commit:
dockertools: Introduce GetPods() for docker runtime.

Commit:
kubelet.container: Add helpers in container runtime to filter pod/containers.

Commit:
kubelet: Replace GetKubeletDockerContainers with GetPods in syncPod/SyncPods.

yifan-gu commented:
Tests fixed. Going to see if it can pass the e2e.

vmarmol commented on pkg/kubelet/container/runtime.go:
```diff
@@ -81,4 +81,34 @@ type Container struct {
+}
+
+// FindContainerByName returns a container in the pod with the given name.
+// Note that when there are multiple containers with the same name, then only
```
nit: We can just say that the first match will be returned, the user can have the data sorted in whatever order they'd like.

xiang90 replied:
+1.

xiang90 commented on pkg/kubelet/container/runtime.go:
```diff
@@ -81,4 +81,34 @@ type Container struct {
+
+type Pods []*Pod
+
+// FindPodByID returns a pod in the pod list by UID. It returns nil
```
The comment does not match the function. It does not return nil when not found.

xiang90 commented on pkg/kubelet/container/runtime.go:
```diff
@@ -81,4 +81,34 @@ type Container struct {
+// FindContainerByName returns a container in the pod with the given name.
+// Note that when there are multiple containers with the same name, then only
+// the newest container with the given name will be return (assuming containers
+// are sorted in reversed time order).
```
It returns nil if not found.

yifan-gu replied:
:(

vmarmol commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -812,3 +813,65 @@ type ContainerCommandRunner interface {
 }
+
+// Parse the pod full name. TODO(yifan): This is duplicated with kubelet.ParsePodFullName.
+func parsePodFullName(podFullName string) (string, string, error) {
```
We can replace other uses of this to keep only 1 version.

yifan-gu replied:
Ye, it causes circular imports. How about putting it in the new runtime.go? @vmarmol

vmarmol replied:
Yes I think so too :) the parsing and unparsing of the podFullName should only happen in one place.

vmarmol commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -812,3 +813,65 @@ type ContainerCommandRunner interface {
+			continue
+		}
+
+		// TODO(yifan): Seems to break container VM users?
```
Container VMs now have UID, no user upgrades Kubelets on those so a new Kubelet will come with new containers.

vmarmol commented on pkg/kubelet/dockertools/docker.go:
```diff
@@ -812,3 +813,65 @@ type ContainerCommandRunner interface {
+		if !found {
+			name, namespace, err := parsePodFullName(podFullName)
+			if err != nil {
+				// TODO(yifan): Or skip the pod?
```
Our current policy is to skip

Commit:
kubelet: Move pod name helpers to pkg/kubelet/container/runtime.go

yifan-gu commented:
@vmarmol I addressed the comments. Moved GodPodFullname and others into the pkg/kubelet/container package.
I think the build is failing because of the master. Waiting for the master to be healthy

Commit:
kubelet.container: Fix volume map type.

yifan-gu commented:
e2e fails 3 tests. I have the same result with HEAD fe65aa5dc98268f0a1a311faa63a272a4e1f16e1 


```
Summarizing 3 Failures:

[Fail] PD [It] should schedule a pod w/ a readonly PD on two hosts, then remove both.
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/pd.go:154

[Fail] Cluster level logging using Elasticsearch [It] should check that logs from pods on all nodes are ingested into Elasticsearch
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/es_cluster_logging.go:337

[Fail] Services [It] should correctly serve identically named services in different namespaces on different external IP addresses
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:372

Ran 29 of 31 Specs in 2204.132 seconds
FAIL! -- 26 Passed | 3 Failed | 2 Pending | 0 Skipped watchEvents output: |

dchen1107 commented on pkg/kubelet/kubelet.go:
```diff
@@ -1580,31 +1584,27 @@ func (kl *Kubelet) SyncPods(allPods []api.Pod, podSyncTypes map[types.UID]metric
-		dockerName, _, err := dockertools.ParseDockerName(dockerContainers[ix].Names[0])
-		_, found := desiredPods[dockerName.PodUID]
-		if err == nil && found {
+	for _, pod := range runningPods {
```
We might misread the logic here, but shouldn't you introduce a bug in a update case described as the below:

 A running pod with two containers: c1 and c2, now update pod with only c1. With your new change, c2 couldn't be removed?

yifan-gu replied:
@dchen1107 I think the syncPod() will take care of the update case.

So the logic I want to express here is:

```
foreach desired pod:
    find its corresponding running pod, and do syncPod.
    
foreach running pod:
    if it is not desired, kill the whole pod.
```

The first loop makes sure we have all the pod up to the desired state. The second loop makes sure we remove any unnecessary pods.

The assumption here is all containers we created can be grouped into a pod.

dchen1107 replied:
ACK.

dchen1107 commented:
LGTM. Could you please rebase again for merge?

yifan-gu commented:
Rebased. Running the e2e again.

vmarmol commented:
LGTM @yifan-gu thanks!

yifan-gu commented:
e2e result:
```
Summarizing 1 Failure:

[Fail] Services [It] should correctly serve identically named services in different namespaces on different external IP addresses
/go/src/github.com/GoogleCloudPlatform/kubernetes/_output/dockerized/go/src/github.com/GoogleCloudPlatform/kubernetes/test/e2e/service.go:372

Ran 29 of 31 Specs in 1447.767 seconds
FAIL! -- 28 Passed | 1 Failed | 2 Pending | 0 Skipped watchEvents output: |
```
I think this is not related with this PR: #5722

yifan-gu commented:
...shippable is failing me again :(

dchen1107 commented:
I kicked out shippable again. we can merge once it is green.

yifan-gu commented:
@dchen1107 Thank you! Since I don't have access to the shippable's host of kubenetes repo, so I am not sure about the reason.
But my personal fork also fails on the master e1a66ad01569006615f6da2aff7c30e5367bfdce.

```
 GOPATH=$PWD/Godeps/_workspace:$GOPATH go install ./...
cluster/addons/dns/kube2sky/kube2sky.go:30:2: cannot find package "github.com/GoogleCloudPlatform/kubernetes/pkg/api" in any of:
	/home/shippable/.gvm/gos/go1.4/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api (from $GOROOT)
	/home/shippable/workspace/src/github.com/yifan-gu/kubernetes/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api (from $GOPATH)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/api
cluster/addons/dns/kube2sky/kube2sky.go:31:2: cannot find package "github.com/GoogleCloudPlatform/kubernetes/pkg/client" in any of:
	/home/shippable/.gvm/gos/go1.4/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client (from $GOROOT)
	/home/shippable/workspace/src/github.com/yifan-gu/kubernetes/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client (from $GOPATH)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/client
cluster/addons/dns/kube2sky/kube2sky.go:32:2: cannot find package "github.com/GoogleCloudPlatform/kubernetes/pkg/fields" in any of:
	/home/shippable/.gvm/gos/go1.4/src/github.com/GoogleCloudPlatform/kubernetes/pkg/fields (from $GOROOT)
	/home/shippable/workspace/src/github.com/yifan-gu/kubernetes/Godeps/_workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/fields (from $GOPATH)
	/home/shippable/workspace/src/github.com/GoogleCloudPlatform/kubernetes/pkg/fields

...

```

dchen1107 commented:
All tests are passed, but unfortunately it looks like you need to rebase again. Sorry about such churns, but good part is that kubelet side changed dramatically everyday. :-)

yifan-gu commented:
@dchen1107 No worries. Just rebased. Wait for green :)

yifan-gu commented:
e2e looks good, only the same test fails.

dchen1107 commented:
Merge now since shippable is green and that e2e test failed on HEAD to me today.

yifan-gu commented:
@dchen1107 Thank you! Next I will try to figure out how to refactor the syncPod.
